Feature/splunkhec

gyrospectre opened the pull request:
Added support for HEC token rotation in the Splunk provider.

Note: Also cleaned up the spec format for this provider, which is a breaking change! Existing Splunk secrets under management will need to be updated with a (now mandatory) config section. This makes things consistent with other providers, so is worth a little extra work.

Commit:
Added HEC support to Splunk provider

Commit:
Added tests

vludax commented on src/keydra/providers/splunk.py:
```diff
@@ -234,20 +294,33 @@ def distribute(self, secret, destination):
+            if spec['config']['type'] not in ['hectoken', 'user']:
+                return (
+                    False,
+                    'Invalid Splunk secret type'.format(
```
It would be very user friendly to specify what the valid types are in the same message 🙂

vludax commented on src/keydra/providers/splunk.py:
```diff
@@ -57,44 +125,36 @@ def _rotate_secret(self, secret):
                 )
             )
 
         return {
```
Hmm we may want to rethink the support for multiple hosts in a single secret. Suppose we have two hosts, A and B:
- When `secrettype == 'user'`, suppose rotation on host A succeeds, but fails on host B. This will raise an error on line 73, and (I think) the secret won't be distributed. Subsequent rotation will fail a the password has been updated on host A, but hasn't been distributed to Secrets Manager.
- When `secrettype == 'hectoken'`... UPD: after reading the validation code, I see we only support a single host

vludax commented on src/keydra/clients/splunk.py:
```diff
@@ -205,6 +210,37 @@ def change_passwd(self, username, oldpasswd, newpasswd):
+        inputs = json.loads(response['body'].read())
+
+        for entry in inputs['entry']:
+            if entry['name'] == 'http://'+inputname:
```
Based on the example in `docsrc/providers.rst`, will `entry['name']` be something like `http://hec1`? 🤔

vludax commented on src/keydra/providers/splunk.py:
```diff
@@ -234,20 +294,33 @@ def distribute(self, secret, destination):
+                        return (False, "'rotatewith' must contain 'provider' and 'key'")
+
+                if len(spec['config']['hosts']) != 1:
+                    return (False, "Only one host can be specified if HEC is the target")
```
Great to see some meaningful validation! 😃

vludax reviewed (approved):
Some food for thought -- perhaps we should consider splitting this up into a separate provider, something like `splunk_hec`. The flows for the two secret types in the current provider appear to have little in common: 

- The validation is largely different
- The rotation process is different
- Multiple hosts is only supported for `user` type
- The tests are separate

One of the few things that are common is `from keydra.clients.splunk import SplunkClient` 😄

Commit:
Moved HEC to a separate provider

gyrospectre commented:
Per feedback, moved HEC to a separate provider, and changed the existing Splunk provider to only allow one rotation host at a time. This will avoid strange failure scenarios.

vludax commented on docsrc/providers.rst:
```diff
@@ -467,3 +468,48 @@ If your app uses storage passwords (like the Qualys app), the distribution stanz
+   key: hec1
+   description: Splunk HEC Rotation Example
+   custodians: your_team
+   provider: splunk
```
Should this say `splunk_hec`?

vludax commented on docsrc/providers.rst:
```diff
@@ -467,3 +468,48 @@ If your app uses storage passwords (like the Qualys app), the distribution stanz
+
+   {
+   "provider": "splunk",
+   "key": "HEC Input Name",
```
I'm a little confused by the `key` here -- is this the input name that we expect to match on `src/keydra/clients/splunk.py:230`? If so, should `src/keydra/providers/splunk_hec.py:73` pass `self._credentials['key']`  (which is part of the secret blob) rather than `secret['key']` (which is the config spec)? Otherwise this `key` field doesn't appear to be used, like `provider`.

I'd consider renaming some of these. E.g. within the secret we could call it something like `hecInputName` instead of `key` to make it unambiguous; and `hecToken` instead of `secret`.

Could also rename the vars within the provider class (`secret` -> `spec`), but that's a smaller concern.

gyrospectre replied:
Fair enough. The `key` from the spec is the one that is used, but I guess I wanted it saved into the SM secret so there's a way to see which input that token relates to from there.

I'll change the SM secret key names to make it more intuitive, and remove the provider key - this is not needed at all.

vludax replied:
> The key from the spec is the one that is used

Would be good to make it clear in the docs somewhere :) Right now it looks like we only refer to "input name" within the docs when describing the contents of the secret. 

Although to me it would make more sense to separate the two -- the spec `key` is used to locate the secret in SM, and I don't believe we use it for any other purposes in other providers. I think drawing that boundary between providers and core Keydra could avoid some confusion.

Commit:
Refactor key names for usability

Commit:
Fixes
